### DISCUSSION TITLE
Drop Shadow Possible?

### DISCUSSION THREAD

**Discussion by Isak-Andersson** [ImageMagick/ImageMagick#3880] [Help]
Hello. Is it possible to add a drop shadow to an image like this?
![Leaf](https://user-images.githubusercontent.com/78758898/123560218-4444d400-d7a1-11eb-89c0-356bc0823c46.png)
![LeafDS](https://user-images.githubusercontent.com/78758898/123560219-44dd6a80-d7a1-11eb-8853-97afc1dfdd69.png)

I've tried this:
`magick.exe convert Leaf.png \( +clone -background black -shadow 100x10+0+0 \) +swap \-layers merge +repage LeafDS.png`

This gives me these errors:

> convert: unable to open image '\(': No such file or directory @ error/blob.c/OpenBlob/3537.
convert: no decode delegate for this image format `' @ error/constitute.c/ReadImage/572.
convert: unable to open image '\)': No such file or directory @ error/blob.c/OpenBlob/3537.
convert: unable to open image '\)': No such file or directory @ error/blob.c/OpenBlob/3537.
convert: no decode delegate for this image format `' @ error/constitute.c/ReadImage/572.
convert: unable to open image '\-layers': No such file or directory @ error/blob.c/OpenBlob/3537.
convert: unable to open image '\-layers': No such file or directory @ error/blob.c/OpenBlob/3537.
convert: no decode delegate for this image format `' @ error/constitute.c/ReadImage/572.
convert: unable to open image 'merge': No such file or directory @ error/blob.c/OpenBlob/3537.
convert: unable to open image 'merge': No such file or directory @ error/blob.c/OpenBlob/3537.
convert: no decode delegate for this image format `' @ error/constitute.c/ReadImage/572.

It creates two output images for some reason. Also, the output is only black with alpha and it expands the canvas which I don't want. Any ideas on how to fix this? I'm on windows 10 if it matters. Thanks!

**Comment by fmw42** [comment]
One way to do it in Unix syntax is:

`magick leaf.png \( +clone -alpha extract -negate -blur 0x10 \) +swap -layers merge +repage result.png`

On Windows, remove the \ from the parentheses

`magick.exe leaf.png ( +clone -alpha extract -negate -blur 0x10 ) +swap -layers merge +repage result.png`

![result](https://user-images.githubusercontent.com/13458218/123561019-daa1d700-d75a-11eb-8072-9533360c2815.png)

In IM 7, use magick, not magick convert.

**Reply by Isak-Andersson** [reply]
Thanks for the quick reply. Is there a way to maintain the alpha channel in the output?

**Reply by fmw42** [reply]
I do not understand. If you put the alpha channel back, it will cover the shadow. So what is it that you want to happen?

**Reply by Isak-Andersson** [reply]
I want the background to be transparent basically. Can it be done?

**Reply by fmw42** [reply]
You are not making sense. You cannot have it transparent and shadowed at the same time.

**Reply by Isak-Andersson** [reply]
Well, I did it in paint.net. Look at the second image I uploaded. Sorry if I wasn't clear enough.

**Reply by fmw42** [reply]
Try this in Unix syntax (change as above for Windows)

`magick leaf.png \( +clone -alpha extract -negate -blur 0x10 -write mpr:shadow \) +swap -layers merge +repage \( mpr:shadow -negate \) -alpha off -compose copy_opacity -composite result.png`

![result](https://user-images.githubusercontent.com/13458218/123562054-13dd4580-d761-11eb-9e7d-f6ce570de659.png)

**Reply by Isak-Andersson** [reply]
It's almost what I want except the Leaf has some unwanted transparency at the edges. Do you know how to fix it?

**Reply by fmw42** [reply]
try

`magick \( leaf.png -channel a -blur 0x2 -level 50x100% +channel \) \( +clone -alpha extract -negate -blur 0x12 -write mpr:shadow \) +swap -layers merge +repage \( mpr:shadow -negate \) -alpha off -compose copy_opacity -composite result.png`

![result](https://user-images.githubusercontent.com/13458218/123562456-abdc2e80-d763-11eb-981c-1dd2da20e89b.png)

**Reply by Isak-Andersson** [reply]
I don't see a big difference, to be honest. It's still transparent at the edges.

**Reply by fmw42** [reply]
I removed the aliasing on the boundary of the leaf.  I guess I do not understand what you want.  Perhaps you should tell us exactly how you processed the image in Paint so we can try to duplicate it.

**Reply by Isak-Andersson** [reply]
I used a plugin for Paint.net called "drop shadow" by Kris Vandermotten.

https://forums.getpaint.net/topic/8375-drop-shadow-and-other-effects-krisvdms-plugin-pack-updated-2010-10-26/

All it does is add a shadow to the image without altering the non-transparent part. So in this case the leaf remains completely opaque. Only the shadow gets progressively more transparent the further away from the edge it is. Hope this makes sense.

**Reply by fmw42** [reply]
That really does not help me to understand how it was created. But try

`magick leaf.png \( +clone -alpha extract -negate -blur 0x20 -negate \) -alpha off -compose copy_opacity -composite result.png`

![result](https://user-images.githubusercontent.com/13458218/123563359-09bf4500-d769-11eb-814a-0467f646efd7.png)

**Reply by Isak-Andersson** [reply]
The shadow itself looks right now but the edges are still transparent.

**Reply by fmw42** [reply]
I do not know what you mean by the "edges are transparent". Please clarify. I thought that was what I fixed that you said did not make any difference. So I am not sure what you mean here.

**Reply by Isak-Andersson** [reply]
It means the Opacity/Alpha value at the edges of the leaf is around 127 (50%) and it should be 255 (100%). In other words, the entire leaf should be 100% opaque. Does that make sense?

**Reply by fmw42** [reply]
Try this:

```
magick leaf.png \
\( +clone -alpha extract +write mpr:alpha \
-negate -fill "gray(20%)" -opaque black -blur 0x20 -negate mpr:alpha \
-compose plus -composite \) \
-alpha off -compose over -compose copy_opacity -composite result.png
```

![result](https://user-images.githubusercontent.com/13458218/123564532-a2a48f00-d76e-11eb-82ea-9d3a242eb36a.png)

For Windows change remove the \ before the parens and change the end of line \ to ^.  To make the shadow lighter/darker increase/decrease the gray(20%) percent amount.

**Reply by Isak-Andersson** [reply]
That's it! Thank you for your help and patience. Much appreciated.
